src/compiler/parser.ts

```diff
@@ -3210,6 +3285,155 @@ module ts {
             node.name = parseRightSideOfDot(/*allowIdentifierNames*/ true);
             return finishNode(node);
         }
+        
+        // Return a string representation of an EntityName
+        function entityNameToString(entityName: EntityName): string {
+            if (entityName.kind === SyntaxKind.Identifier) {
+                return (<Identifier>entityName).text;
+            } else {
+                let qualifiedName = <QualifiedName>entityName;
+                return entityNameToString(qualifiedName.left) + '.' + entityNameToString(qualifiedName.right);
+            }
+            return "";
+        }
+
+        function parseJsxChild(): JsxText | JsxExpression | JsxElement {
+            let result: JsxText | JsxExpression | JsxElement = undefined;
+            switch (token) {
+                case SyntaxKind.JsxText:
+                    result = <JsxText>createNode(SyntaxKind.JsxText);
+                    token = scanner.scanJsxToken();
+                    result = finishNode(result);
+                    break;
+                case SyntaxKind.OpenBraceToken:
+                    result = parseJsxExpression();
+                    break;
+                default:
+                    Debug.assert(token === SyntaxKind.LessThanToken);
+                    result = parseJsxElement();
+                    break;
+            }
+            token = scanner.reScanJsxToken();
+            Debug.assert(result !== undefined, "parsed some JSX child");
+            return result;
+        }
+
+        function parseJsxElement(): JsxElement {
+            let node = <JsxElement>createNode(SyntaxKind.JsxElement);
+            node.openingElement = parseJsxOpeningElement();
+
+            let tagName = entityNameToString(node.openingElement.tagName);
+
+            if (node.openingElement.isSelfClosing) {
+                node.children = <NodeArray<JsxText | JsxExpression | JsxElement>>[];
+            }
+            else {
+                // Rescan since parsing the > messed up the scanner state
+                token = scanner.reScanJsxToken();
+                node.children = parseList(ParsingContext.JsxChildren, false, parseJsxChild);
+                node.closingElement = parseJsxClosingElement(tagName);
+            } 
+
+            return finishNode(node);
+        }
+        
+        function parseJsxOpeningElement(): JsxOpeningElement {
+            let node = <JsxOpeningElement>createNode(SyntaxKind.JsxOpeningElement);
+
+            parseExpected(SyntaxKind.LessThanToken);
+            node.tagName = parseJsxElementName();
+            node.attributes = parseList(ParsingContext.JsxAttributes, /*checkForStrictMode*/ false, parseJsxAttribute);
+            if (token === SyntaxKind.SlashToken) {

```

Comment:
I would make this a different type of node.  And have this function return a union type of the two types of JsxElement and JsxEmptyElement.